### ISSUE TITLE
Check if Skywire can be run on Android

### ISSUE THREAD

**Issue by nkryuchkov** [skycoin/skywire#99]
The main aim is to allow to port the Skywire Visor and the skysocks to Android with as little modifications to the codebase as possible because it would be great if we did not have to maintain two separate codebases. (apart from the frontend which we could get a designer or even scrap that part initially for a beta). It would be preferable to have an apk which we could distribute without relying on other applications  to be setup. However, if we need to (i.e if an apk would require substantial changes to Skywire itself), we could also release a version that requires other programs such as a shell to be setup.

For simplicity, I think we should get rid of the Hypervisor, dmsgpty and Skychat as well as the peering-daemon for this project. 

The UI could be simply a button for connect/disconnect, a Field for Setting a Pubkey as remote for the proxy and the ability to call the proxy discovery for a server in a certain region(client for which is still to be done, but kifen is on it afaik.)

One issue which would persist (as it exists on Linux as well for now) is interoperability with the Android Browser. We would likely have to configure the browser to use the proxy or set it in the settings of the phone. How other VPNs or proxies do that on Android rn would have to be investigated; I do not have a clue. We could reserve and deploy Skywire visors with non-auth protected proxy servers running on servers throughout different regions for ease of usage initially.
So open questions I see rn:
1. How to get communication between apps and visor working?
2. How to get the browser and other applications on Android to use the proxy connection?
3. How to distribute the software.

_Originally posted by @jdknives in Discord_

**Comment by nkryuchkov** [comment]
### TL;DR

> Check if Skywire can be run on Android 

Yes, it can be.

> How to get communication between apps and visor working?

Add TCP support along with Unix sockets for communication between apps and visor.

> How to get the browser and other applications on Android to use the proxy connection?

Use VPN API instead of adding proxy settings.

> How to distribute the software.

Wrap `skywire-visor` binary in APK.

---

### Description

I've successfully run `skywire-visor` on a phone with Android 9 via `adb shell` without `hypervisor`, `dmsgpty` and `apps`. However, the running process is not as straightforward as on desktop OS. 

First of all, `sdcard` is mounted as non-executable and it's not possible to run a binary from there. However, the `/data/local/tmp` worked great. 

Since `skywire-visor` needs to create some files, file paths in the config should point to a folder with write permissions, e.g. `/data/local/tmp`. Moreover, since `/` is the home folder, it's forbidden to create a `.skycoin` folder there. We need to override the `HOME` environment variable to a folder with write permissions.

I noticed that binding to a Unix socket doesn't work on Android, so we'll probably need to add a network type to configs to support both Unix sockets and TCP.

So assuming the above I think we can wrap the `skywire-visor` binary in an Android app, running visor and managing its config from the app. This link looks helpful:
https://developer.android.com/reference/java/lang/Runtime

It looks like we can manage configs in APK and pass them to binary, and also control the stdout of binary to have logs of what's going on there. However, some further investigation will be helpful here.

There's still a question of how to use Skywire from Android. Adding proxy settings to the Android/Chrome browser seems impossible and not user-friendly. There are VPN APIs on both Android and iOS that are usually used for apps like VPNs (e.g. OpenVPN). So I think this is the only way we can go to make usage of Skywire user-friendly.

Here are the APIs:
1. Android:
- https://developer.android.com/guide/topics/connectivity/vpn
- https://developer.android.com/reference/android/net/VpnService

2. iOS
- https://developer.apple.com/documentation/networkextension

We'll probably need to implement a TUN/TAP adapter for the described above (https://github.com/SkycoinProject/skywire-mainnet/issues/100), however, this needs further investigation.

If we use the VPN API we probably can get rid of running apps like a proxy client.

**Comment by jdknives** [comment]
Great work. Will make a ticket for the mplementation and keep it in the backlog.
